Change mechanism to provide routes

alexpeachey opened the pull request:
@kevgo 
Address #4 by changing the way we specify routes.

The `--route` and `-r` flag are gone and we now expect either the name of a file to use the only non-flag argument or we expect the routes to come from standard in.

We treat routes from stdin slightly differently depending on if there is a TTY or not.

Because we can't know the format without an extension and have no extension when reading from stdin, we have new flags `--format` and `-f` which can be `cson` or `json` and default to `json` if not provided and can't be determined from the extension of the file (if provided)

I added a feature test for reading from stdin, but don't know how to write one to test the alternative path which is used when std in is a TTY.

kevgo commented on features/running_from_piped_input.feature:
```diff
@@ -0,0 +1,43 @@
+        response: "world"
+      ]
+      """
+    When I pipe the file "hello_world.cson" into "psapi --format=cson"
```
Here we pipe data directly into a process. We shouldn't need a separate config file here. All you do in your step definition is reading from it anyways.

```cucumber
Given I pipe the following data into "psapi --format=cson"
  """
  [
    method: "get"
    path: "/hello"
    status: 200
    response: "world"
  ]
  """

alexpeachey replied:
Good point. Can just hide the fact that there is a file.

kevgo commented on features/running_from_piped_input.feature:
```diff
@@ -0,0 +1,43 @@
+  So that I can have another program generate my config.
+
+
+  Scenario: CSON configuration file
```
Scenario: piping CSON data into psapi

kevgo commented on features/step_definitions/run_steps.coffee:
```diff
@@ -11,6 +13,17 @@ module.exports = ->
+    @process = spawn command,
+                     args,
+                     cwd: @tmpDir
+                     stdio: [
```
Use the string captured from Cucumber directly here.

alexpeachey replied:
That works? OK, cool, I thought I had to give it a file descriptor (i.e. open a file)

kevgo replied:
I haven't tried it. But if you need a stream, you could make a string stream or something like that directly from your string. http://stackoverflow.com/questions/12755997/how-to-create-streams-from-string-in-node-js

kevgo commented on features/step_definitions/run_steps.coffee:
```diff
@@ -1,4 +1,6 @@
 spawn = require 'cross-spawn'
+fs = require 'fs'
+path = require 'path'
```
These two dependencies can go

kevgo commented on features/step_definitions/run_steps.coffee:
```diff
@@ -11,6 +13,17 @@ module.exports = ->
+    command = @psapiPath if command is 'psapi'
+    @process = spawn command,
+                     args,
+                     cwd: @tmpDir
```
I like separating arguments onto newlines to make it visually more clear what the different arguments are. If you agree with that logic, a better way of formatting this now is:

```coffeescript
@process = spawn command,
                 args,
                 cwd: @tmpDir, stdio: config_data
```

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
+  catch e
+    console.log 'Could not parse routes.'
+    console.log e
+    process.exit -2
```
Line 14-21 should be its own function that returns the parsed content, or exits the script.

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
-  process.exit -1
+startServer = (err, content) ->
+  if err
+    console.log 'Could not read routes.'
```
This error message seems off. If starting the server fails, shouldn't the error message be something like "Cannot start the server"?

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
-  console.log e
-  process.exit -2
+routesFromTerminal = (done) ->
+  Readline = require 'readline'
```
Move the require to the top of the file, and sort these requires alphabetically there.

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
+  readline = Readline.createInterface
+    input: process.stdin
+    output: process.stdout
+  readline.write 'Enter your config and hit ^D when finished:\n'
```
I don't see a use case where somebody types in the config manually here. I think we only have to support (a) reading from a config file, and (b) piping in a programmatically created config. If somebody wants to type in config, they should do that into a text file.

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
+  content = ''
+  process.stdin.on 'data', (data) ->
+    content = content + data
+  process.stdin.on 'end', -> done null, content
```
This could get extracted into an eventEmitterReader helper.

```coffeescript
routesFromPipe = (done) ->
  process.stdin.resume()
  process.stdin.setEncoding 'utf8'
  eventEmitterReader process.stdin, (err, result) -> 
    done result
```

This helper listens on `data` events, concatenates all data packets, and when it receives `end`, calls the given callback with the concatenated data.

kevgo commented on lib/psapi_cli.coffee:
```diff
@@ -4,29 +4,50 @@ CSON = require 'CSON'
+else if process.stdin.isTTY
+  routesFromTerminal startServer
+else
+  routesFromPipe startServer
```
I think this file has a `Configuration` class waiting to be extracted. The class would be given the command-line arguments in the constructor, and then read the remaining configuration from either a file or stdIn, then make the whole configuration available in a standardized format through fields.

kevgo commented:
Nice functionality! It looks like you are still heavily prototyping here. Sorry if I'm too premature in pushing for a more factored solution here.

alexpeachey commented:
@kevgo Rearranged things quite a bit, removed the explicit interactive mode (though it still works, just doesn't have a separate path and output of instructions) and extracted a `Configuration` class.

kevgo commented on features/exit_codes.feature:
```diff
@@ -5,16 +5,10 @@ Feature: Exit codes
-    Then I see "Could not read routes file"
-    And the command exits with status 2
+    When I run "psapi zonk.coffee"
+    Then I see "Could not read routes."
```
In my opinion specs shouldn't specify punctuation of terminal output. Those are more stylistic UI sugar on top of the basic functionality

kevgo commented on lib/configuration.coffee:
```diff
@@ -0,0 +1,55 @@
+class Configuration
+
+  constructor: ->
+    argv = minimist process.argv.slice 2
```
In order to make this class better testable, I'd prefer if we provided all dependencies as parameters here, including the command-line arguments.

kevgo commented on lib/configuration.coffee:
```diff
@@ -0,0 +1,55 @@
+    argv = minimist process.argv.slice 2
+    @port = argv.port || argv.p || 5000
+    @routesFile = argv._[0]
+    @format = @_formatFromFile(@routesFile) || argv.format || argv.f || 'json'
```
This `||` syntax is really cool for using minimist together with default values!

kevgo commented on lib/event_emitter_reader.coffee:
```diff
@@ -0,0 +1,6 @@
+
```
Remove this empty line

kevgo commented on lib/event_emitter_reader.coffee:
```diff
@@ -0,0 +1,6 @@
+
+module.exports = (stream, done) ->
+  content = ''
+  finished = false
```
Use `no` here, it's more correct  :)

kevgo commented on lib/psapi.coffee:
```diff
@@ -16,8 +16,8 @@ class Psapi
-    console.log "Listening at #{chalk.cyan 'http://localhost:'}#{chalk.cyan port}"
-    @app.listen port
+    @app.listen port, ->
+      console.log "Listening at #{chalk.cyan 'http://localhost:'}#{chalk.cyan port}"
```
Nice!

kevgo commented:
Cool! LGTM after the final comments are addressed.
